### DISCUSSION TITLE
Struct member values

### DISCUSSION THREAD

**Discussion by charles-esterbrook** [electronstudio/raylib-python-cffi#120] [Ideas]
Printing raylib structs gives opaque strings like `<cdata 'struct Color' owning 4 bytes>` whether using `str()` or `repr()`.

The type is there, but not the member values. Whether labeled or not, they would be really useful.

Any chance for that?

**Comment by electronstudio** [comment]
You could probably format this a bit nicer but here is an example:

```
import pyray as pr

camera = pr.Camera3D([18.0, 16.0, 18.0], [0.0, 0.0, 0.0], [0.0, 1.0, 0.0], 45.0, 0)


def dump(struct):
    s = f"{pr.ffi.typeof(struct)}: ("
    for field in dir(struct):
        data = struct.__getattribute__(field)
        if str(data).startswith("<cdata"):
            data = dump(data)
        s = s + f"{field}:{data} "
    s = s + ")"
    return s


print(dump(camera))
```

**Comment by charles-esterbrook** [comment]
Nice. Here is a version which matches Python instantiation syntax more closely:
```python
def structToStr(struct: object) -> str:
    # example output: 'Vector2(x=3.0, y=4.0)'
    name = pr.ffi.typeof(struct).cname.split(' ', 1)[-1]
    parts = [name, "("]
    sep = ''
    for field in dir(struct):
        data = struct.__getattribute__(field)
        dataStr = str(data)
        if dataStr.startswith("<cdata"):
            dataStr = structToStr(data)
        parts.append(sep)
        parts.append(f"{field}={dataStr}")
        sep = ', '
    parts.append(')')
    return ''.join(parts)
```
Any way this could become the default __repr__() of the various types?

If there are objects with too many fields, a `keys` argument could be used to select a subset. Just thinking out loud.

**Reply by electronstudio** [reply]
The cffi types are immutable and can't be monkey patched so I don't see an easy way of replacing repr().  Could perhaps create a proxy object and use that in place of the real object everywhere.  But really this should be done by cffi, maybe ask them.

**Comment by charles-esterbrook** [comment]
I tried making a Vector2 wrapper class with the hooks to make it "tuple-like" via `__len__`, `__getitem__` and `__iter__`, but when passed to raylib functions, they error with:
```
  File "", line 22, in render
    pr.draw_rectangle_v(self.loc, self.size, self.color)
  File "python3.12/site-packages/pyray/__init__.py", line 65, in func
    result = a(*modified_args)
             ^^^^^^^^^^^^^^^^^
TypeError: initializer for ctype 'struct Vector2' must be a list or tuple or dict or struct-cdata, not Vector2
```
Looks like specific types are required rather than "list-like" or "tuple-like". Is that something that can be relaxed?

I have other wrapper ideas in mind such as various methods and operator overloads.

**Reply by electronstudio** [reply]
Even if you make a wrapper that proxies every attribute it seems the cffi can still tell it's not a real cdata object.  So then you would have to write code to detect your wrapper objects and unwrap them on every function call, and re-wrap every result returned back.  It's possible, but the main motivation for the cffi binding was to keep things simple and let cffi do everything automatically.  It's grown a *little* more complex over the years but wrapping/unwrapping everything would seem to be a *lot* of additional complexity.  At some point it wouldn't be worth using cffi at all, you may as well just code everything manually like other bindings do.  I could be wrong of course, there could be some easy way to do it I'm not aware of.

**Comment by charles-esterbrook** [comment]
If the cffi checked for `hasattr` on `__len__, __getitem__` instead of `isinstance(x, list)` for accepting alternate argument types, then wrappers would work the same as lists (`[3, 4]`) and tuples (`(3, 4)`). I haven't yet looked into where the checks for the list and tuple are being specified and/or generated.

**Reply by electronstudio** [reply]
I’m not sure how much CFFI/ctypes are even Python objects. I think this stuff is implemented in C as part of CPython itself, so it’s not going to be easy to mess with from the Python side. 

Also, something that has come up in the past, different Python implementations do CFFI/ctypes differently, so even if you get it working on CPython you might find it doesn’t work on PyPy.

**Comment by electronstudio** [comment]
We found a way of making a vector2 wrapper that seems to work in place of the real vector2: https://github.com/electronstudio/raylib-python-cffi/blob/master/examples/extra/vector2_extended.py
